README
		lomoco - Logitech Mouse Control for USB mice

	            http://lomoco.linux-gamers.net/

  Copyright (C) 2004 Alexios Chouchoulas
  Copyright (C) 2005 Andreas Schneider, Peter Feuerer, Tobias Schleuss

  lomoco is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


All comments are welcome.


ChangeLog
lomoco (1.0)

  * Update udev rules an script for latest version
  * Added patch for the MX518, MX510, MX310, MX3100
  * Added lomoco.xml manpage template
  * Updated udev support, thanks to Ludwig Nussel
  * Fixed memory leaks

 -- Andreas Schneider <andreas.schneider@linux-gamers.net>  Wed Mar 01 15:05:38 CEST 2006


autogen.sh
#!/bin/sh
#
# lomoco autogen script
#

# Automakeversion
AM_1=1
AM_2=7
AM_3=6

# Autoconfversion
AC_1=2
AC_2=59

# Check automake version
AM_VERSION=`automake --version | sed -n -e 's#[^0-9]* \([0-9]*\)\.\([0-9]*\)\.*\([0-9]*\).*$#\1 \2 \3#p'`
AM_V1=`echo $AM_VERSION | awk '{print $1}'`
AM_V2=`echo $AM_VERSION | awk '{print $2}'`
AM_V3=`echo $AM_VERSION | awk '{print $3}'`

if [ "$AM_1" -gt "$AM_V1" ]; then
	AM_ERROR=1 
else
	if [ "$AM_1" -eq "$AM_V1" ]; then
		if [ "$AM_2" -gt "$AM_V2" ]; then
			AM_ERROR=1 
		else
			if [ "$AM_2" -eq "$AM_V2" ]; then
				if [ -n "$AM_V3" -o "$AM_3" -gt "$AM_V3" ]; then
					AM_ERROR=1 
				fi
			fi
		fi
	fi
fi

if [ "$AM_ERROR" = "1" ]; then
	echo -e  '\E[31;m'
	echo -n "Your automake version `automake --version | sed -n -e 's#[^0-9]* \([0-9]*\.[0-9]*\.[0-9]*\).*#\1#p'`"
	echo " is older than the suggested one, $AM_1.$AM_2.$AM_3"
	echo "Go on at your own risk. :-)"
	echo
	tput sgr0
fi

# Check autoconf version
AC_VERSION=`autoconf --version | sed -n -e 's#[^0-9]* \([0-9]*\)\.\([0-9]*\).*$#\1 \2#p'`
AC_V1=`echo $AC_VERSION | awk '{print $1}'`
AC_V2=`echo $AC_VERSION | awk '{print $2}'`

if [ "$AC_1" -gt "$AC_V1" ]; then
	AC_ERROR=1 
else
	if [ "$AC_1" -eq "$AC_V1" ]; then
		if [ "$AC_2" -gt "$AC_V2" ]; then
			AC_ERROR=1 
		fi
	fi
fi

if [ "$AC_ERROR" = "1" ]; then
	echo -e  '\E[31;m'
	echo -n "Your autoconf version `autoconf --version | sed -n -e 's#[^0-9]* \([0-9]*\.[0-9]*\).*#\1#p'`"
	echo " is older than the suggested one, $AC_1.$AC_2"
	echo "Go on at your own risk. :-)"
	echo
	tput sgr0
fi

if [ "$LT_ERROR" = "1" ]; then
	echo -e  '\E[31;m'
	echo -n "Your libtool version `libtool --version | sed -n -e 's#[^0-9]* \([0-9]*\.[0-9]*\).*#\1#p'`"
	echo " is older than the suggested one, $LT_1.$LT_2"
	echo "Go on at your own risk. :-)"
	echo
	tput sgr0
fi

echo Configuring build environment for lomoco
echo ...
aclocal \
  && autoheader --force \
  && automake --add-missing --foreign --copy \
  && autoconf --force \
  && echo "Now running configure to configure lomoco" \
  && echo "./configure $@" \
  && ./configure $@


src/g_hack.c
/*
 * gcc -o g_hack g_hack.c
 * ./g_hack /dev/hidraw0
 *
 * g_hack - proof of conecpt code
 *
 *
 * Copyright (c) 2006-2009 Andreas Schneider <mail@cynapses.org>
 * Copyright (c) 2006-2007 Peter Feuerer <piie@piie.net>
 *
 * License: GPLv2 or later
 *
 * Additional Info:
 *
 * the G-series has two modes:
 *
 * 1. mode: no driver
 * in this mode there is a list of resolutions implemented in the mouse
 * and you can switch between them using the buttons
 *
 * 2. mode is: windows driver
 * in this mode the buttons talk to the driver
 
 * they have absolute no effect on the hardware in first place
 * the driver gets the button event looks in its table, which resolution to use
 * sends it to the mouse sets the leds and the resolution
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <asm/types.h>
#include <fcntl.h>
#include <unistd.h>
#include <linux/hiddev.h>
 
#define VERSION "0.0.7"
 
#define VENDOR 0x046d
#define MOUSE_G5 0xc041
#define MOUSE_G7 0xc51a
 
// Icehawk78 - first insertion for new mouse product code
#define MOUSE_MX1100 0xc526
 
/* thanks to Michael "sanni" Dewein */
#define MOUSE_G3 0xc042
 
/* thanks to Niels "aboe" Abspoel */
#define MOUSE_G9 0xc048
 
 
static int send_report(int fd, const unsigned char *buf, size_t size) {
  struct hiddev_report_info rinfo;
  struct hiddev_usage_ref uref;
  int i, err;
 
  for (i = 0; i < size; i++) {
    memset(&uref, 0, sizeof(uref));
    uref.report_type = HID_REPORT_TYPE_OUTPUT;
    uref.report_id   = 0x10;
    uref.field_index = 0;
    uref.usage_index = i;
    uref.usage_code  = 0xff000001;
    uref.value       = buf[i];
 
    err = ioctl(fd, HIDIOCSUSAGE, &uref);
 
    if (err < 0)
      return err;
  }
 
  memset(&rinfo, 0, sizeof(rinfo));
  rinfo.report_type = HID_REPORT_TYPE_OUTPUT;
  rinfo.report_id   = 0x10;
  rinfo.num_fields  = 1;
  err = ioctl(fd, HIDIOCSREPORT, &rinfo);
 
  return err;
}
 
void send_msg(int fd,char c0,char c1,char c2, char c3, char c4, char c5)
{
  unsigned char b[6];
  b[0]=c0;
  b[1]=c1;
  b[2]=c2;
  b[3]=c3;
  b[4]=c4;
  b[5]=c5;
 
  if(send_report(fd,b,6)<0)
  {
    perror("error sending to device");
    close(fd);
    exit(1);
  }
}
int main (int argc, char **argv) {
 
  int fd = -1;
  int version;
  int resolution=0;
  struct hiddev_devinfo device_info;
 
  printf("g5hack version: %s\n\n", VERSION);
 
  /* ioctl() requires a file descriptor, so we check we got one, and
     then open it */
  if (argc != 3) {
    fprintf(stderr, "Usage: %s hiddevice resolution\n - hiddevice probably /dev/usb/hiddev0 or /dev/hidraw0\n - resolution: 0=400 1=800 2=1600 3=2000\n"
        "only g9: 4-7 (7=3200)\n\n", argv[0]);
 
    exit(1);
  }
  if ((fd = open(argv[1], O_RDONLY)) < 0) {
    perror("hiddev open");
    exit(1);
  }
  resolution=atoi(argv[2]);
 
  /* ioctl() accesses the underlying driver */
  ioctl(fd, HIDIOCGVERSION, &version);
 
  /* the HIDIOCGVERSION ioctl() returns an int
   * so we unpack it and display it
   * we create a patch
   */
  printf("hiddev driver version is %d.%d.%d\n", version >> 16, (version >> 8) & 0xff, version & 0xff);
 
  /* suck out some device information */
  ioctl(fd, HIDIOCGDEVINFO, &device_info);
 
  /* the HIDIOCGDEVINFO ioctl() returns hiddev_devinfo
 
   * structure - see <linux/hiddev.h>
   * So we work through the various elements, displaying
   * each of them
   */
  printf("vendor 0x%04hx product 0x%04hx version 0x%04hx ",
      device_info.vendor, device_info.product,
      device_info.version);
  printf("has %i application%s ", device_info.num_applications,
      (device_info.num_applications==1?"":"s"));
  printf("and is on bus: %d   devnum: %d   ifnum: %d\n",
      device_info.busnum, device_info.devnum,
      device_info.ifnum);
 
// Icehawk78 - Second section - need to add in the check so that the product continues with your mouse code. The printf isn't necessary, and was mostly added for consistency sake.
  /* We have a G5? */
  if((device_info.vendor == (short)VENDOR) &&
      ((device_info.product == (short)MOUSE_MX1100) ||
       (device_info.product == (short)MOUSE_G3) ||
       (device_info.product == (short)MOUSE_G9) ||
       (device_info.product == (short)MOUSE_G5) ||
       (device_info.product == (short)MOUSE_G7))) {
 
    if(device_info.product == (short)MOUSE_MX1100)
      printf(">> MX1100 Mouse detected!\n");
 
    if(device_info.product == (short)MOUSE_G3)
      printf(">>  G3 Gaming Mouse detected!\n");
 
    if(device_info.product == (short)MOUSE_G9)
      printf(">>  G9 Gaming Mouse detected!\n");
 
    if(device_info.product == (short)MOUSE_G5)
      printf(">>  G5 Gaming Mouse detected!\n");
 
    if(device_info.product == (short)MOUSE_G7)
      printf(">>  G7 cordless Gaming Mouse detected!\n");
    /*
     * Initialise the internal report structures
     *
     * The G5 has an additional HID Interface
     * with a report descriptor.
     */
    if (ioctl(fd, HIDIOCINITREPORT, 0) < 0) {
      perror("hid report init failed");
      exit(1);
 
    }
 
/* Icehawk78 - Now we're getting into the meat of it.
   This is the first place that you'll have to change.
   On my mouse, when changing between any two non-DPI-related modes,
   I captured a total of three URB_CONTROL out signals. The first and the third matched
   the first four digits [10 01 80 65] while the second did not [10 01 80 63].
   The second signal appears to be the first required signal, which I added here as
   a super-hacky conditional. Note that you drop the first 0x10 because send_msg()
   is already including that, for some reason. */
 
    /* disable speed buttons */
    if (device_info.product != (short)MOUSE_MX1100) {
      send_msg(fd,0x00, 0x80, 0x01, 0x00, 0x00, 0x00);
    } else {
      send_msg(fd,0x01,0x80,0x63,0x82,0x00,0x00);
    }
 
/* Icehawk78 - This is the second location to change.
   The first/third capture codes I put into the hacky if block inside
   case 0 and case 1; case 0 enables the buttons, case 1 sets it back to
   DPI mode. */
 
    switch(resolution){
      case 0:
        if (device_info.product == (short)MOUSE_MX1100) {
          send_msg(fd,0x01,0x80,0x65,0x82,0x85,0xFF);
          break;
        } else {
          /* set current speed to 400dpi */
          send_msg(fd,0x00,0x80,0x63,0x80,0x00,0x00);
 
          /* turn off all speed leds */
          send_msg(fd,0x00,0x80,0x51,0x11,0x01,0x00);
          break;
        }
 
      case 1:
        if (device_info.product != (short)MOUSE_MX1100) {
          /* set current speed to 800dpi */
          send_msg(fd,0x00,0x80,0x63,0x81,0x00,0x00);
          /* turn on first led */
          send_msg(fd,0x00,0x80,0x51,0x12,0x01,0x00);
          break;
        } else {
          send_msg(fd,0x00,0x80,0x65,0x82,0x85,0x9A);
          break;
        }
 
      case 2:
        /* set current speed to 1600dpi */
        send_msg(fd,0x00, 0x80, 0x63, 0x82, 0x00, 0x00);
        /* turn on second led */
        send_msg(fd,0x00, 0x80, 0x51, 0x21, 0x01, 0x00);
 
        break;
 
      case 3:
        /* set current speed to 2000dpi */
        send_msg(fd,0x00, 0x80, 0x63, 0x83, 0x00, 0x00);
        /* turn on third led */
        send_msg(fd,0x00, 0x80, 0x51, 0x11, 0x02, 0x00);
        break;
 
      case 4:
        /* set current speed to whatever dpi
         * - only available for g9 */
        if(device_info.product == (short)MOUSE_G9){
          send_msg(fd,0x00, 0x80, 0x63, 0x84, 0x00, 0x00);
          /* turn on third led */
          send_msg(fd,0x00, 0x80, 0x51, 0x11, 0x02, 0x00);
        }
        else{
          printf("speed not supported for your device\n");
        }
        break;
 
      case 5:
        /* set current speed to whatever dpi
         * - only available for g9 */
        if(device_info.product == (short)MOUSE_G9){
          send_msg(fd,0x00, 0x80, 0x63, 0x85, 0x00, 0x00);
          /* turn on third led */
          send_msg(fd,0x00, 0x80, 0x51, 0x11, 0x02, 0x00);
        }
        else{
          printf("speed not supported for your device\n");
        }
        break;
 
      case 6:
        /* set current speed to whatever dpi (most probably 3200)
         * - only available for g9 */
        if(device_info.product == (short)MOUSE_G9){
          send_msg(fd,0x00, 0x80, 0x63, 0x86, 0x00, 0x00);
          /* turn on third led */
          send_msg(fd,0x00, 0x80, 0x51, 0x11, 0x02, 0x00);
        }
        else{
 
          printf("speed not supported for your device\n");
        }
        break;
 
      case 7:
        /* set current speed to whatever dpi (most probably 3200)
         * - only available for g9 */
        if(device_info.product == (short)MOUSE_G9){
          send_msg(fd,0x00, 0x80, 0x63, 0x87, 0x00, 0x00);
          /* turn on third led */
          send_msg(fd,0x00, 0x80, 0x51, 0x11, 0x02, 0x00);
        }
        else{
          printf("speed not supported for your device\n");
        }
        break;
 
      default:
        printf("speed not supported for your device\n");
        break;
    }
 
    /* foobar: */
    /* at lowest speed ("-" button does not function anymore) */
    /* send_msg(fd,0x00, 0x81, 0x51, 0x00, 0x00, 0x00); */
 
    /* turn off all speed leds */
    /* send_msg(fd,0x00, 0x80, 0x51, 0x11, 0x01, 0x00); */
 
    /* turn on all speed leds */
    /* send_msg(fd,0x00, 0x80, 0x51, 0x22, 0x02, 0x00); */
 
  } /* G5 */
 
  close(fd);
 
  exit(0);
}

src/lomoco.h
/* 
   Formerly lmctl.h - Main header file.
   Copyright (C) 2004 Alexios Chouchoulas

   lomoco.h - Main header file

   Copyright (C) 2005 Andreas Schneider <andreas.schneider@linux-gamers.net>
                      Peter Feuerer <peter.feuerer@linux-gamers.net>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  
*/


#ifndef __LOMOCO_H
#define __LOMOCO_H

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <stropts.h>
#include <string.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/select.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <pty.h>
#include <iconv.h>
#include <fcntl.h>
#include <unistd.h>


#include <usb.h>

#include "cmdline.h"



#define __KERNEL__
        
#define VENDOR_LOGITECH 0x046d
#define TIMEOUT         0x1000      
                        
                        
#define P6  buf [0]
#define P0  buf [1]
#define P4  buf [2]
#define P5  buf [3]
#define P8  buf [4]
#define P9  buf [5]
#define PB0 buf [6]     
#define PB1 buf [7]

                
#define max(a,b) ((a) > (b) ? (a) : (b))

/* for debugging perror */
#define perror printf("%s:%d\n",__FILE__,__LINE__);perror

/* program name (lomoco) */
extern char *progname;

/*
 * pid       Product id
 * name      Mouse name
 * model     Mouse Model
 *
 * has_csr   Mouse with a receiver
 * has_res   Resolution control
 * has_ssr   Smart Control Reporting
 * has_sms   Smart Control
 *
 * is_dual   Dual receiver, mouse and keyboard
 */
typedef struct {
            ushort      pid;
/*@null@*/  char  *name;
/*@null@*/  char  *model;

            int   has_csr;
            int   has_res;
            int   has_ssr;
            int   has_sms;

            int   is_dual;
} mouse_t;

#endif /* __LOMOCO_H */

src/cmdline.h
/* 
   cmdline.h - Command-line parser interface.

   Copyright (C) 2002 Alexios Chouchoulas
   Copyright (C) 2005 Andreas Schneider <andreas.schneider@linux-gamers.net>
                      Peter Feuerer <peter.feuerer@linux-gamers.net>


   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  
*/


#ifndef __CMDLINE_H
#define __CMDLINE_H

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif
#include <argp.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "lomoco.h"

#define _(x) x

/* argp option keys */

typedef enum {
      cmd_none = 0,
      cmd_scan,
      cmd_set,
      cmd_inquire,
} command_t;


enum {
      DUMMY_KEY=129,
      KEY_SMS,
      KEY_NO_SMS
};

/* Option flags and variables.  These are initialized in parse_opt.  */

extern command_t   command;
extern int         set_res;
extern int         set_sms;
extern char       *match_bus;
extern char       *match_dev;
extern int         match_pid;
extern int         set_channel;
extern int         set_lock;
extern int         set_unlock;
extern int         set_clear;

void cmdline_parse (int argc, char **argv);

#endif /* __CMDLINE_H */

cmdline.c
/* 
   cmdline.c - Command-line parser.

   Copyright (C) 2004 Alexios Chouchoulas
   Copyright (C) 2005 Andreas Schneider <andreas.schneider@linux-gamers.net>
                      Peter Feuerer <peter.feuerer@linux-gamers.net>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

*/

#include "cmdline.h"

/* Option flags and variables. These are initialized in parse_opt. */

command_t command     = cmd_none;
int       set_res     = 0;
int       set_sms     = 0;
int       set_channel = 0;
int       set_lock    = 0;
int       set_unlock  = 0;
int       set_clear   = 0;

/*@null@*/char    *match_bus = NULL;
/*@null@*/char    *match_dev = NULL;
/*@null@*/int     match_pid = 0x0000;

/*@null@*/char    *local=NULL;            /* --local=LOCAL-ENCODING */
/*@null@*/char    *remote=NULL;           /* --remote=REMOTE-ENCODING */
/*@null@*/char    **cmdline=NULL;         /* SHELL-COMMAND */


/* prototypes */
static error_t parse_opt (int key, char *arg, struct argp_state *state);
static void show_version (FILE *stream, struct argp_state *state);


/* The argp functions examine these global variables.  */
const char *argp_program_bug_address = "<lomoco-dev@lists.linux-gamers.net>";
void (*argp_program_version_hook) (FILE *, struct argp_state *) = show_version;

static char args_doc[] = "";

/* This is the description table for all options argp needs to handle for lomoco */
struct /*@null@*/argp_option options[] =
{
      { NULL, 0, NULL, 0, _("Device Matching"), 0 },

      { "bus",         'b',     "USB-BUS-NUMBER",             0,
        "Only apply the command to the specified bus, specified in "
        "full (e.g. 001 for bus 1)", 0},
      { "dev",         'd',     "USB-DEV-NUMBER",             0,
        "Only apply the command to the specified device number, specified "
        "in full (e.g. 034 for device 34)", 0},
      { "pid",         'p',     "PRODUCT-ID",             0,
        "Only apply the command to the specified product ID, specified "
        "in hexadecimal and in full (e.g. C025 for an MX-500 optical mouse)", 0},

      { NULL, 0, NULL, 0, _("Actions"), 0 },

      { "scan",        's',     NULL,             0,
        "Scan the USB bus and report any matching, supported Logitech mice", 0},
      { "inquire",     'i',     NULL,             0,
        "Obtain and display the status of all supported capabilities "
        "for all matching devices", 0},

      { NULL, 0, NULL, 0, _("Wireless Status Reporting (CSR command set)"), 0 },

      { "ch1",         '1',     NULL,             0,
        "Set matching devices to channel 1", 0},
      { "ch2",         '2',     NULL,             0,
        "Set matching devices to channel 2", 0},
      { "unlock",      'u',     NULL,             0,
        "Unlock the receiver and search for a wireless mouse in range.", 0},
      { "lock",        'l',     NULL,             0,
        "Stop the 30 second search for a wireless mouse.", 0},
      { "clear",       'c',     NULL,             0,
        "Make the receiver forget about the mouse connected to it.", 0},

      { NULL, 0, NULL, 0, _("Resolution Changing (RES command set)"), 0 },

      { "400",         '4',     NULL,             0,
        "Set matching devices to a resolution of 400cpi", 0},
      { "800",         '8',     NULL,             0,
        "Set matching devices to a resolution of 800cpi", 0},
      { "1200",        'm',     NULL,             0,
        "Set matching devices to a medium resolution of 1200cpi", 0},
      { "1600",        'h',     NULL,             0,
        "Set matching devices to a high resolution of 1600cpi", 0},

      { "2000",        'g',     NULL,             0,
        "Set matching devices to a high resolution of 2000cpi", 0},

      { NULL, 0, NULL, 0, _("SmartScroll/Cruise Control (SMS command set)"), 0 },

      { "sms",         KEY_SMS, NULL,             0,
        "Enable SmartScroll/Cruise Control. 'Up' and 'down' buttons also issue"
        "wheel up/down events", 0},
      { "no-sms",      KEY_NO_SMS,     NULL,             0,
        "Disable SmartScroll/Cruise Control. 'Up' and 'down' buttons do "
        "not also issue wheel events", 0},
      
      { NULL, 0, NULL, 0, NULL, 0 }
};

struct /*@null@*/argp argp =
{
      options, parse_opt, args_doc,
      "Manage vendor-specific features of Logitech USB mice.",
      NULL, NULL, NULL
};





/*
 * Descripton: Show the version number and copyright information.
 *
 * Parameters:
 *
 * Return:
 */
static void show_version (FILE *stream, struct argp_state *state)
{
      /* 
       * Print in small parts whose localizations can hopefully be copied
       * from other programs.
       */
      fputs(PACKAGE" "VERSION"\n", stream);
      fprintf(stream, 
            "fork of lmctl written by Alexios Chouchoulas <alexios@bedroomlan.org>.\n"
            "Copyright (C) %s %s\n"
            "Copyright (C) %s %s\n"
            "This program is free software; you may redistribute it under the terms of\n"
            "the GNU General Public License.  This program has absolutely no warranty.\n"
            "\n",
            "2003", "Alexios Chouchoulas",
            "2005", "Andreas Schneider, Peter Feuerer");
}


/*
 * Description: Parse a single option. Is called by argp_parse and performes steps for
 * every single option.
 *
 * Parameters: key (which option was entered) 
 *             *arg (string after option, e.g. -b 001 <-- *arg would contain 001)
 *             *state (TODO) 
 *
 * Return: 
 */
static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
      int dummy;

      switch (key) {
      case 'b':
            match_bus = strdup (arg);
            if ((sscanf (arg, "%03d", &dummy) != 1) ||
                (strlen (arg) != 3) ||
                (strspn (arg, "0123456789") != 3)) {
                  fprintf (stderr,
                         "%s: the bus ID must be a three-digit decimal\n"
                         "number (e.g. 005). Use lsusb(1) to see the proper"
                         "format. you must specify the\nUse %s --help for"
                         "usage information.\n",
                         progname, progname);
                  exit (1);
            }
            break;
            
      case 'd':
            match_dev = strdup (arg);
            if ((sscanf (arg, "%03d", &dummy) != 1) ||
                (strlen (arg) != 3) ||
                (strspn (arg, "0123456789") != 3)) {
                  fprintf (stderr,
                         "%s: the device ID must be a three-digit decimal\n"
                         "number (e.g. 005). Use lsusb(1) to see the proper"
                         "format. you must specify the\nUse %s --help for"
                         "usage information.\n",
                         progname, progname);
                  exit (1);
            }
            break;
            
      case 'p':
            if ((sscanf (arg, "%x", &match_pid) != 1) ||
                (match_pid > 0xffff) || (match_pid < 0)) {
                  fprintf (stderr,
                         "%s: the product ID must be a hexadecimal "
                         "number (e.g. c025),\nup to four digits long. "
                         "Use lsusb(1) to see the proper format. "
                         "you must\nspecify the\nUse %s --help for "
                         "usage information.\n",
                         progname, progname);
                  exit (1);
            }
            break;
            
      case 's':               /* --scan */
            command = cmd_scan;
            break;

      case 'i':               /* --inquiry */
            command = cmd_inquire;
            break;

      case '4':         /* set the resolution */
      case '8':
            command = cmd_set;
            set_res = (int) (key - '0') * 100;
            break;
      
      case 'm':
            command = cmd_set;
            set_res = (int) (12) * 100;
            break;
      case 'h':
            command = cmd_set;
            set_res = (int) (16) *100;
            break;
      case 'g':
            command = cmd_set;
            set_res = (int) (20) *100;
            break;


      case '1':         /* set the channel */
      case '2':
            command = cmd_set;
            set_channel = (int) (key - '0');
            break;
      
      case 'u':
            command = cmd_set;
            set_unlock = 1;
            break;

      case 'l':
            command = cmd_set;
            set_lock = 1;
            break;

      case 'c':
            command = cmd_set;
            set_clear = 1;
            break;

      case KEY_SMS:
      case KEY_NO_SMS:  /* enable/disable SMS */
            command = cmd_set;
            set_sms = key == KEY_SMS? 1: -1;
            break;

      case ARGP_KEY_ARG:
            /* End processing here. */
            cmdline = &state->argv [state->next - 1];
            state->next = state->argc;
            break;

      default:
            return ARGP_ERR_UNKNOWN;
      }
      
      return 0;
}

/*
 * Description: gets called by main function of lomoco.c
 *              and starts the argument parsing function 
 * 
 * Parameters:  argc (sum of args)
 *              **argv (array of args)
 *
 * Return:      nothing, exits program with (EXIT_FAILURE) if something failes
 */

void cmdline_parse (int argc, char **argv)
{
      /* Parse the arguments */
      
      argp_parse (&argp, argc, argv, ARGP_IN_ORDER, NULL, NULL);

      /* Validate the arguments */

      if (command == cmd_none) {
            fprintf (stderr,
                   "%s: you must specify a command.\n"
                   "\nUse %s --help for usage information.\n",
                   progname, progname);
            exit (EXIT_FAILURE);
            
      }
      
      /* Set up the default command, if one is needed. */
}

